### PR TITLE
Mark the panic_no_unwind lang item as nounwind

### PR DESCRIPTION
**Pull request by Amanieu**
This has 2 effects:
- It helps LLVM when inlining since it doesn't need to generate landing pads for `panic_no_unwind`.
- It makes it sound for a panic handler to unwind even if `PanicInfo::can_unwind` returns true. This will simply cause another panic once the unwind tries to go past the `panic_no_unwind` lang item. Eventually this will cause a stack overflow, which is safe.

### CODE REVIEW

**Comment by Amanieu** [comment]
cc @nbdd0121

**Comment by lcnr** [comment]
r? @nagisa maybe

don't know enough here

File: compiler/rustc_typeck/src/collect.rs [hunk 1]
```diff
@@ -2778,6 +2778,13 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
         }
     }
 
+    // The panic_no_unwind function called by TerminatorKind::Abort will never
+    // unwind. If the panic handler that it invokes unwind then it will simply
+    // call the panic handler again.
+    if Some(id) == tcx.lang_items().panic_no_unwind() {
+        codegen_fn_attrs.flags |= CodegenFnAttrFlags::NEVER_UNWIND;
```
Comment (nbdd0121) [inline compiler/rustc_typeck/src/collect.rs:2785]: Could we use `#[rustc_allocator_nounwind]` instead? There's nothing allocator-specific for that attribute even though it has allocator in its name.

Comment (nagisa) [reply compiler/rustc_typeck/src/collect.rs:2785]: If we did this, I would suggest renaming the attribute.

**Comment by nagisa** [comment]
@bors r+

I think its okay if we rely on language items to figure out what attributes we want to add to specific ones since we do that already quite pervasively. I can see us using a dedicated attribute once we have one. Right now what we have is pretty arbitrarily scoped to specific component of the standard library and I'm not entirely sure its a responsibility of this PR to adapt it to be more general.

@Amanieu your description mentions soundness improvements. Do you think this is something that should appear in the relnotes?

**Comment by Amanieu** [comment]
This is only reachable with nounwind functions (`extern "C"` with `#![feature(c_unwind)]`, `Drop` with `-Z panic-in-drop=abort`, `#[rustc_allocator_nounwind]`) all of which are currently unstable. I don't think this merits appearing in the relnotes.

File: compiler/rustc_typeck/src/collect.rs (hunk 1 above)
Comment (Amanieu) [reply compiler/rustc_typeck/src/collect.rs:2785]: I agree that this should be an attribute (or possibly a `extern "Rust-nounwind"` ABI?). I've been away this past week and haven't had time to implement this yet, but it is on my TODO list.

**Pull request merged**